lib/HLSL/DxilModule.cpp

```diff
@@ -367,7 +389,65 @@ void DxilModule::CollectShaderFlags(ShaderFlags &Flags) {
           case DXIL::OpCode::Msad:
             hasMSAD = true;
             break;
-          case DXIL::OpCode::BufferLoad:
+          case DXIL::OpCode::BufferLoad: {
+              Value *resHandle = CI->getArgOperand(DXIL::OperandIndex::kBufferStoreHandleOpIdx);
+              CallInst *handleCall = cast<CallInst>(resHandle);
+
+              if (ConstantInt *resClassArg =
+                  dyn_cast<ConstantInt>(handleCall->getArgOperand(
+                      DXIL::OperandIndex::kCreateHandleResClassOpIdx))) {
+                  DXIL::ResourceClass resClass = static_cast<DXIL::ResourceClass>(
+                      resClassArg->getLimitedValue());
+                  if (resClass == DXIL::ResourceClass::UAV) {
+                    if (ConstantInt *resRangeID =
+                            dyn_cast<ConstantInt>(handleCall->getArgOperand(
+                                DXIL::OperandIndex::kCreateHandleResIDOpIdx))) {
+                      DxilResource &res = GetUAV(resRangeID->getLimitedValue());
+                      if (res.GetKind() == DXIL::ResourceKind::TypedBuffer) {
+                        Value *GV = res.GetGlobalSymbol();
+                        llvm::Type *Ty = GV->getType()->getPointerElementType();
```

Comment:
Does this work correctly for indexable resources?  Should we be using GetRetType() instead?  The main missing piece is unwrapping outer-most arrays.

On that note, I'm not sure GetRetType() is 100% correct either, it would probably be good to review that as well.